### PR TITLE
Fix problem with deprecated array_key_exists() for object.

### LINKED ISSUE #14: Deprecated: array_key_exists(): in RouteCollectorFactory.php on line 59
### Bug Report

|    Q        |   A
|------------ | ------
| Version(s)  | 3.3.0

#### Summary

Deprecated: array_key_exists(): Using array_key_exists() on objects is deprecated. Use isset() or property_exists() instead in /app/vendor/mezzio/mezzio-router/src/RouteCollectorFactory.php on line 59

#### Current behavior
изображение

#### How to reproduce
Use latest version of symfony components (symfony di, etc)

#### Expected behavior

without warnings. 

I propose replace array_key_exists -> isset or use type hint.
https://github.com/mezzio/mezzio-router/blob/3.4.x/src/RouteCollectorFactory.php#L59

### PR DESCRIPTION
**Pull request by nepster-web**
Add ArrayObject support for config.

More details in issue https://github.com/mezzio/mezzio-router/issues/14

### CODE REVIEW

File: src/RouteCollectorFactory.php [hunk 1]
```diff
@@ -55,7 +55,7 @@ private function getDetectDuplicatesFlag(ContainerInterface $container): bool
             return true;
         }
 
-        $config = $container->get('config');
+        $config = (array)$container->get('config');
```
Comment (samsonasik) [inline src/RouteCollectorFactory.php]: probably better with instanceof `ArrayObject` check, eg:

```suggestion
        $config = $container->get('config');
        $config = $config instanceof \ArrayObject
            ? $config->getArrayCopy()
            : $config;
```

Comment (nepster-web) [reply src/RouteCollectorFactory.php]: If we support a common interface, the implementation may be different. 
so, maybe other implementations will implement custom collections, like: MyArrayWrapper or ArrayIterator and etc.

If we use `$config instanceof \ArrayObject` we will close only 2 case: with simple array and ArrayObject.

Comment (Ocramius) [reply src/RouteCollectorFactory.php]: Array casts on objects are not usage of object public API, and expose object internals in unintentional ways. Operating with the `ArrayAccess` API (and checking for that type) is more correct here.

**Comment by samsonasik** [comment]
there is no method `getArrayCopy` in `ArrayAccess`, also, please provide unit test

File: src/RouteCollectorFactory.php (hunk 1 above)
Comment (nepster-web) [reply src/RouteCollectorFactory.php]: @Ocramius, I fixed its like @samsonasik proposed, however with ArrayAccess.

**Comment by nepster-web** [comment]
@samsonasik, fair. working on it

**Review by Ocramius — APPROVED** [review]
LGTM.

Just an additional note: since this is a bugfix, let's target `3.3.x` instead of `3.4.x`

**Comment by nepster-web** [comment]
@samsonasik 
https://github.com/mezzio/mezzio-router/blob/050022a8e0982a2dec0347f484bd5639fd7fded3/test/RouteCollectorFactoryTest.php#L90

https://github.com/mezzio/mezzio-router/blob/050022a8e0982a2dec0347f484bd5639fd7fded3/src/RouteCollectorFactory.php#L61

We solve a specific problem with ArrayObject. as you suggested.

File: src/RouteCollectorFactory.php [hunk 2]
```diff
@@ -56,6 +57,11 @@ private function getDetectDuplicatesFlag(ContainerInterface $container): bool
         }
 
         $config = $container->get('config');
+
+        $config = $config instanceof ArrayObject
```
Comment (Ocramius) [inline src/RouteCollectorFactory.php]: Ah, just noticed that we check for `ArrayObject` specifically: IMO better to check for `ArrayAccess` and then use `offsetExists` explicitly, but not really that attached to that opinion too strongly.

Comment (nepster-web) [reply src/RouteCollectorFactory.php]: I think this will require more changes for getDetectDuplicatesFlag method.

**Comment by samsonasik** [comment]
please apply DCO https://github.com/mezzio/mezzio-router/pull/15/checks?check_run_id=1668060328

**Comment by nepster-web** [comment]
@samsonasik, DCO interesting thing. I didn't know about it.

**Comment by nepster-web** [comment]
@samsonasik, what do you think about the release?

**Comment by samsonasik** [comment]
There is 1 travis build failure https://travis-ci.com/github/mezzio/mezzio-router/jobs/469882033

could you verify? Thank you.

**Comment by nepster-web** [comment]
@samsonasik, @Ocramius  check it please. 
The analyzer doesn't like code duplication and doesn't understand data types.

**Review by samsonasik — APPROVED** [review]

**Comment by guilhermeaiolfi** [comment]
Shouldn't this be merged and released?

File: src/RouteCollectorFactory.php [hunk 3]
```diff
@@ -56,12 +60,21 @@ private function getDetectDuplicatesFlag(ContainerInterface $container): bool
         }
 
         $config = $container->get('config');
-        if (! array_key_exists(RouteCollector::class, $config)) {
+
+        $config = is_array($config) ? new ArrayObject($config) : $config;
```
Comment (Ocramius) [inline src/RouteCollectorFactory.php:64]: IMO should be simplified by removing the need for `ArrayObject` at all: we can `assert($config instanceof ArrayAccess || is_array($config))` instead.

File: src/RouteCollectorFactory.php [hunk 4]
```diff
@@ -56,12 +60,21 @@ private function getDetectDuplicatesFlag(ContainerInterface $container): bool
...
+
+        if (! $config instanceof ArrayAccess) {
```
Comment (Ocramius) [inline src/RouteCollectorFactory.php:66]: Can be dropped based on assertion

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Review by nepster-web — COMMENTED** [review]
good too

**Pull request merged**